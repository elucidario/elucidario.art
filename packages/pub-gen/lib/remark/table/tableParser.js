import path from "path";
import { visit } from "unist-util-visit";
import { parseNodeValue, isPubGenNodeValue } from "../../utils.js";
import { getPaths, readFile } from "@elucidario/pkg-paths";
import * as unist from "@elucidario/pkg-unist";
import * as tableschema from "tableschema";
import { SchemaValidator } from "@elucidario/pkg-schema-validator";
import { Console } from "@elucidario/pkg-console";

function tableParser(treeOptions) {
    const { publication, lang, assets, version, pkg, ...extra } = treeOptions;

    const console = new Console(pkg);

    // default docxProcessor to pandoc while we don't implement remark-docx
    let docxProcessor = "pandoc";
    if (typeof extra.docxProcessor !== "undefined")
        docxProcessor = extra.docxProcessor;

    const schemaValidator = new SchemaValidator();

    const tableSchema = readFile(
        path.resolve(
            getPaths().packages,
            "pub-gen",
            "static",
            "pub-gen",
            "schemas",
            "table-schema.json",
        ),
    ).content;

    return (tree, file) => {
        visit(tree, "text", async (node) => {
            if (!isPubGenNodeValue(node.value)) return;

            const { action, filePath, fileOptions, options, type } =
                parseNodeValue(node.value);

            if (action !== "table") return;

            const tableFile = readFile(
                path.resolve(
                    getPaths().publications,
                    publication,
                    "content",
                    lang,
                    filePath,
                ),
            ).content;

            const valid = schemaValidator.validate({
                schema: tableSchema,
                data: tableFile,
            });

            if (!valid) {
                console.error(schemaValidator.getErrors());
                throw schemaValidator.getErrors();
            }

            const { data, title, note, fields } = tableFile;

            const tableAst = [
                title ? unist.bold([unist.text(title)]) : false,
                title ? unist.breakNode() : false,
                title ? unist.breakNode() : false,
                unist.table(
                    "left",
                    data.map((row) => {
                        return unist.tableRow(
                            row.map((cell) => {
                                return unist.tableCell([unist.text(cell)]);
                            }),
                        );
                    }),
                ),
                note ? unist.breakNode() : false,
                note ? unist.breakNode() : false,
                note
                    ? unist.paragraph([
                          unist.bold([unist.text(note.label)]),
                          unist.text(": "),
                          unist.text(note.content),
                      ])
                    : false,
                unist.breakNode(),
                unist.breakNode(),
            ].filter(Boolean);

            node.type = "paragraph";
            node.children = tableAst;

            delete node.value;
        });
    };
}

export default tableParser;
