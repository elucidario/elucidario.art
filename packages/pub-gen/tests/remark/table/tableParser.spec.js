import path from "path";
import tableParser from "../../../lib/remark/table/tableParser";
import { readFile, getPaths } from "@elucidario/pkg-paths";
import testProcessor from "../../mocks/remarkProcessor";

describe("tableParser", () => {
    const tableContent = readFile(
        path.resolve(
            getPaths().publications,
            "publicacao-teste",
            "content",
            "pt-br",
            "internal",
            "body",
            "really",
            "first-level.md",
        ),
    ).value;

    it("should return tableParser node", async () => {
        const tableParserNode = await testProcessor(tableContent, {
            plugin: tableParser,
            pluginOptions: {
                publication: "publicacao-teste",
                lang: "pt-BR",
            },
            stringifyOptions: {},
            publication: "publicacao-teste",
            lang: "pt-BR",
        });

        expect(tableParserNode).toMatchSnapshot();
    });
});
