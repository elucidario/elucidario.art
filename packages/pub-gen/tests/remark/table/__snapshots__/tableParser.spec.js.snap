// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tableParser should return tableParser node 1`] = `
VFile {
  "cwd": "C:\\Users\\55119\\Elucid√°rio.art\\elucidario\\packages\\pub-gen",
  "data": {},
  "history": [],
  "messages": [],
  "value": "# First level

{{count:quadros;legend=Teste Tabela.}}

| Name       | ID |
| :--------- | -- |
| John Doe   | 1  |
| John Smith | 3  |

Fonte: Elaborado pelo autor.
",
}
`;
